Linter fixes pybindings

ll-nick opened the pull request:
Here comes another PR regarding linter issues analogous to #138 and #145.

### TL;DR
This PR adjusts the python bindings to the changes introduced in #138 and turns the Python package into a fully typed library.

### Details

The first few commits concern adjustments for breaking changes introduced in #138. As mentioned before, binding `util_caching` is no longer necessary and was removed in 49cd944eaeb1222239657f09ab86d2f243bd74f5.

Next, I dove down deep into the rabbit whole of static type checking in Python. I opted to use pyright which appears to be the de-facto standard in this regard. So the newer commits concern adding typing annotations to all classes/methods defined in the unit test which already fixes some of the issues with pyright.

To fully utilize pyright, I had to go a step further than that though since it does not fully understand the types of the arbitration graph library due to it being C++ bindings and due to the `EnvironmentModelWrapper` and `CommandWrapper` classes which do not have corresponding classes on the Python side of things (which is done on purpose to abstract away the complexities involved in binding a template C++ library).

However, there is a solution to such issues: [stub files](https://mypy.readthedocs.io/en/stable/stubs.html), see also the [PEP 561](https://peps.python.org/pep-0561/).

> A stub file is a file containing a skeleton of the public interface of that Python module, including classes, variables, functions – and most importantly, their types.

Creating those can be a little bit cumbersome but fortunately [`pybind11-stubgen`](https://github.com/sizmailov/pybind11-stubgen) helped me get a good starting point (see bd1a50d02b625d18ee7de2381fb5160fb3be24c0) which I then refined in the following commits.
The stub files (ending in `*.pyi`) are then installed together with a `py.typed` marker file indicating that this is a typed library (see [here](https://peps.python.org/pep-0561/#packaging-type-information) for reference).

With these files, static type checkers such as pyright can do their magic and warn you if you implement things you'll regret at runtime :)
Given the type checks I also found some smaller issues with the existing bindings (6323ce78d1f279e4d9f7eeb1a9ae5a69e223cd0c, 82d337de78de8df7ea9607b3f4f72ccc11f45d0b, e0b8c856b95f52aa8a76a5571562cf5dcfae88ec) and unit tests (da74f1e08f9173159a27e7b9717207429805b1a0)

In summary, the library is now fully typed while all issues that pyright had with the unit tests are now fixed as well.

ll-nick committed:
Use new getter functions for option.behavior and behavior.name

ll-nick committed:
Bind (relevant) new getter function

ll-nick committed:
Adjust for changed flag style case

ll-nick committed:
It's now toString (or to_string in python)

ll-nick committed:
Access verification result via new getter function

ll-nick committed:
Fix import ordering and remove unused imports

ll-nick committed:
Add init py for tests

ll-nick committed:
No need to bind util_caching anymore :)

ll-nick committed:
Replace wild card import and sys.modules hack with extensive list of modules

ll-nick committed:
Do not export flag enum into option class

ll-nick committed:
Add missing named argument

ll-nick committed:
Fix implicit string concatination

ll-nick committed:
Turn off some checks in the unit tests

It is common practice to use the setUp function to define test fixtures.
No need to declare them in the constructor.

ll-nick committed:
Mark test classes as final to fix warnings about potential type changes

For reference, see [reportUnannotatedClassAttribute]

ll-nick committed:
Mark functions as override rather than suppressing warning

ll-nick committed:
Bind estimate_cost function

ll-nick committed:
Add typing annotations in python implemented classes

ll-nick committed:
Add autogenerated pyi stubs to enable static type checking

Generated via `pybind11-stubgen arbitration_graphs`

ll-nick committed:
Add a typing module for some aliases that improve semantics

ll-nick committed:
Add py.typed marker file according to PEP 561

See https://peps.python.org/pep-0561/#packaging-type-information

ll-nick committed:
Fix verification import

ll-nick committed:
Remove unused import

ll-nick committed:
Use custom type aliases in stub files

ll-nick committed:
Fix return type of options()

ll-nick committed:
Do not use else after return

ll-nick committed:
Do not report usage of any in the typing stubs

ll-nick committed:
Be explicit about overrides in stub files as well

ll-nick committed:
Suppress some warnings that cannot be resolved

ll-nick committed:
Suppress warnings regarding non-inherited Flags type

ll-nick committed:
Fix warning regarding wrong override of add_option

ll-nick committed:
Expect DummyCommand rather than string in cost estimator cost map

ll-nick committed:
Add types for the YAML representation of an arbitration graph

ll-nick committed:
Add some explanatory comments as why the some type aliases exist

ll-nick committed:
Explicitely unwrap optional verification_result before accessing is_ok()

ll-nick committed:
Fix a couple more implicit string concatinations

ll-nick committed:
Fix a couple missing type hints in random arbitrator test

ll-nick committed:
Fix typing issues regarding ambiguous node type in yaml representation

ll-nick committed:
Remove fixed version string from stub file and import version from shared library

ll-nick commented on python_bindings/src/arbitration_graphs/__init__.pyi:
```diff
@@ -0,0 +1,377 @@
+
+class CostArbitrator(Arbitrator):
+    class Option(Arbitrator.Option):
+        # The nested "Flags" class intentionally differs between Arbitrator subtypes.
+        # Each Arbitrator subclass defines its own Option.Flags type even though pyright
+        # expects the child class to override the Flags enum of the parent class.
+        # We therefore suppress the pyright warning.
+        class Flags:  # pyright: ignore[reportIncompatibleVariableOverride]
```
This right here is actually a valid warning. However, we cannot fix that on the Python side since it's caused by how the `Flags` enum is defined on C++ side.

orzechow replied:
Yeah… I'm not happy with the implementation of those flags as well.
For now, let's keep it as is (as of our time constraints), but keep it in mind for future improvements 🤞

ll-nick committed:
Make typing compatible with Python 3.8

Support for the previously used syntax was added in later Python version

orzechow reviewed (approved):
This is huge – python static type checking is great!

Only downside: yet another layer of maintenance overhead.
Could you please add docs on how to keep these bindings and typings up to date (in a future PR)? 

I guess we could collect all these topics in a [contributors guideline](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors): #147

orzechow commented:
Build fails, because we need to first merge #145 and rebase this one on top the updated update-interfaces branch.

orzechow commented:
Rebased on top of `update-interfaces` with #145 merged (c08afc5)

ll-nick committed:
Run unit tests from pybind root directory

Running tests from inside the test directory prevent python from
recognizing it as a package causing the relative imports to break

orzechow commented:
51c9a55 – awesome @ll-nick!
